src/datafusion/table_factory.rs

```diff
@@ -0,0 +1,106 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use datafusion::{
+    datasource::{listing::ListingTableUrl, provider::TableProviderFactory, TableProvider},
+    error::DataFusionError,
+    execution::context::SessionState,
+    logical_expr::CreateExternalTable,
+};
+
+use super::table_provider::{ListingZarrTableConfig, ListingZarrTableOptions, ZarrTableProvider};
+
+struct ZarrListingTableFactory {}
+
+#[async_trait]
+impl TableProviderFactory for ZarrListingTableFactory {
+    async fn create(
+        &self,
+        state: &SessionState,
+        cmd: &CreateExternalTable,
+    ) -> datafusion::common::Result<Arc<dyn TableProvider>> {
+        if cmd.file_type != "ZARR" {
+            return Err(datafusion::error::DataFusionError::Execution(
+                "Invalid file type".to_string(),
+            ));
+        }
+
+        let table_path = ListingTableUrl::parse(&cmd.location)?;
+
+        let options = ListingZarrTableOptions {};
+        let schema = options
+            .infer_schema(state, &table_path)
+            .await
+            .map_err(|e| DataFusionError::Execution(format!("infer error: {:?}", e)))?;
+
+        let table_provider =
+            ZarrTableProvider::new(ListingZarrTableConfig::new(table_path), schema);
+
+        Ok(Arc::new(table_provider))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use datafusion::execution::{
+        config::SessionConfig,
+        context::{SessionContext, SessionState},
+        runtime_env::RuntimeEnv,
+    };
+
+    use crate::tests::get_test_v2_data_path;
+
+    #[tokio::test]
+    async fn test_create() -> Result<(), Box<dyn std::error::Error>> {
```

Comment:
This test shows how this looks in SQL.

Reply:
Nice, so with this we can apply any SQL operation (that datafusion supports) to a zarr store? Also, and this isn't necessarily surprising if that's the case, I had hoped it would work like that, but it seems we could, for example, create a separate table off of some parquet data, and it to the session, and then everything could be seamlessly combined via a query?

Reply:
Yeah, exactly. To your point, you could register a parquet-based table then do a join with it, or write the output of a zarr query to parquet, etc.